lib/services/unified_resource.go

```diff
@@ -128,31 +129,27 @@ func (cfg *UnifiedResourceCacheConfig) CheckAndSetDefaults() error {
 func (c *UnifiedResourceCache) putLocked(resource resource) {
 	key := resourceKey(resource)
 	sortKey := makeResourceSortKey(resource)
-	oldResource, exists := c.resources[key]
-	if exists {
+	if old, exists := c.resources[key]; exists {
 		// If the resource has changed in such a way that the sort keys
 		// for the nameTree or typeTree change, remove the old entries
 		// from those trees before adding a new one. This can happen
 		// when a node's hostname changes
-		oldSortKey := makeResourceSortKey(oldResource)
-		if oldSortKey.byName.Compare(sortKey.byName) != 0 {
-			c.deleteSortKey(oldSortKey)
+		if old.sortKey.byName.Compare(sortKey.byName) != 0 {
+			c.deleteSortKey(old.sortKey)
 		}
+		old.sortKey = sortKey
 	}
-	c.resources[key] = resource
+	if c.resources[key] == nil {
+		c.resources[key] = newResourceCollection(resource.GetKind(), sortKey)
+	}
+	c.resources[key].put(resource)
```

Comment:
nit: maybe we can reduce number of `c.resources[key]` here, currently it does it 3-4 times per `putLocked`.